pkg/execute/mapping.go

```diff
@@ -0,0 +1,150 @@
+	// noFeature is used for commands that have no features defined
+	noFeature = ""
+)
+
+// CommandVerb are commands supported by the bot
+type CommandVerb string
+
+// CommandVerb command options
+const (
+	CommandPing     CommandVerb = "ping"
+	CommandHelp     CommandVerb = "help"
+	CommandVersion  CommandVerb = "version"
+	CommandFeedback CommandVerb = "feedback"
+	CommandList     CommandVerb = "list"
+	CommandEnable   CommandVerb = "enable"
+	CommandDisable  CommandVerb = "disable"
+	CommandEdit     CommandVerb = "edit"
+	CommandStart    CommandVerb = "start"
+	CommandStop     CommandVerb = "stop"
+	CommandStatus   CommandVerb = "status"
+	CommandConfig   CommandVerb = "config"
+)
+
+// CommandExecutor defines command structure for executors
+type CommandExecutor interface {
+	Commands() map[CommandVerb]CommandFn
+	FeatureName() FeatureName
+}
+
+// CommandFn is a single command (eg. List())
+type CommandFn func(ctx context.Context, cmdCtx CommandContext) (interactive.Message, error)
+
+// CommandContext contains the context for CommandFn
+type CommandContext struct {
+	Args            []string
+	ClusterName     string
+	CommGroupName   string
+	BotName         string
+	RawCmd          string
+	User            string
+	Conversation    Conversation
+	Platform        config.CommPlatformIntegration
+	ExecutorFilter  executorFilter
+	NotifierHandler NotifierHandler
+	Mapping         *CommandMapping
+}
+
+// FeatureName defines the name and aliases for a feature
+type FeatureName struct {
+	Name    string
+	Aliases []string
+}
+
+// CommandMapping allows to register and lookup commands and dynamically build help messages
+type CommandMapping struct {
+	commands map[CommandVerb]map[string]CommandFn
+	help     map[CommandVerb][]FeatureName
+}
+
+// NewCmdsMapping registers command and help mappings
+func NewCmdsMapping(executors []CommandExecutor) (*CommandMapping, error) {
+	mappingsErrs := multierror.New()
+	cmdsMapping := make(map[CommandVerb]map[string]CommandFn)
+	helpMapping := make(map[CommandVerb][]FeatureName)
+	for _, executor := range executors {
+		cmds := executor.Commands()
+		subCmd := executor.FeatureName()
+		for verb, cmdFn := range cmds {
+			if value := cmdsMapping[verb]; value == nil {
+				cmdsMapping[verb] = make(map[string]CommandFn)
+			}
+			if value := helpMapping[verb]; value == nil {
+				helpMapping[verb] = make([]FeatureName, 0)
+			}
+			cmdsMapping[verb][subCmd.Name] = cmdFn
+			helpMapping[verb] = append(helpMapping[verb], subCmd)
+			for _, featureName := range subCmd.Aliases {
+				if _, ok := cmdsMapping[verb][featureName]; ok {
+					mappingsErrs = multierror.Append(mappingsErrs, fmt.Errorf("command collision: tried to register '%s %s', but it already exists", verb, featureName))
+				}
+				cmdsMapping[verb][featureName] = cmdFn
+			}
+		}
+	}
+	mapping := &CommandMapping{
+		commands: cmdsMapping,
+		help:     helpMapping,
+	}
+	return mapping, mappingsErrs.ErrorOrNil()
```

Comment:
**nit** In Go there is a pattern to return the zero value for other values when err occurs. It's also more readable as you see the err paths directly. It's also consistent with other principle to align the happy path to the left edge.

```suggestion
	if err := mappingsErrs.ErrorOrNil(); err != nil {
		return nil, err
	}

	return &CommandMapping{
		commands: cmdsMapping,
		help:     helpMapping,
	}, nil
```